syntax = "proto3";

package distributedstore;

service KeyValueStore {
  rpc put(PutRequest) returns (PutResponse);

  rpc get(GetRequest) returns (GetResponse);

  rpc slowDown(SlowDownRequest) returns (SlowDownResponse);

  rpc restore(RestoreRequest) returns (RestoreResponse);

  rpc registerNode(RegisterRequest) returns (Empty);

  rpc canCommit(VoteRequest) returns (VoteResponse);

  rpc doCommit(CommitRequest) returns (CommitResponse);

  rpc doAbort(AbortRequest) returns (Empty);

  rpc gossip(GossipRequest) returns (Empty);

  rpc askVote(VoteRequest) returns (VoteResponse);

}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
  bool found = 2;
}

message SlowDownRequest {
  int32 seconds = 1;
}

message SlowDownResponse {
  bool success = 1;
}

message RestoreRequest {}

message RestoreResponse {
  bool success = 1;
}

message RegisterRequest {
  string address = 1;
  int32 weight = 2;
}

message RegisterResponse {
  bool success = 1;
}

message VoteRequest {
  string key = 1;
}

message VoteResponse {
  bool vote = 1;
  string value = 2;
}

message CommitRequest {
  string key = 1;
  string value = 2;
}

message CommitResponse {
  bool committed = 1;
}

message AbortRequest {
  string key = 1;
}

message GossipRequest {
  string address = 1;
  int32 weight = 2;
}

message Empty {
    
}